independent-projects/resteasy-reactive/server/processor/src/main/java/org/jboss/resteasy/reactive/server/processor/generation/filters/CustomFilterGenerator.java

```diff
@@ -466,12 +476,58 @@ private static FieldDescriptor generateConstructorAndDelegateField(ClassCreator
         ResultHandle delegate = ctor.getMethodParam(0);
         ctor.writeInstanceField(delegateField, self, delegate);
         ctor.returnValue(null);
+
+        if (needsToBeApplicationScoped) {
+            // generate no-args constructor needed for creating proxies
+            MethodCreator noArgsCtor = cc.getMethodCreator("<init>", void.class);
+            noArgsCtor.setModifiers(Modifier.PUBLIC);
+            noArgsCtor.invokeSpecialMethod(MethodDescriptor.ofConstructor(superClassName), noArgsCtor.getThis());
+            noArgsCtor.writeInstanceField(delegateField, noArgsCtor.getThis(), noArgsCtor.loadNull());
+            noArgsCtor.returnValue(null);
+        }
+
         return delegateField;
     }
 
-    private static ResultHandle[] getResponseFilterResultHandles(MethodInfo targetMethod, DotName declaringClassName,
-            MethodCreator filterMethod, int filterMethodParamCount, ResultHandle rrReqCtxHandle,
-            Set<DotName> unwrappableTypes) {
+    /**
+     * The generated class needs to be {@code @ApplicationScoped} if one of the following is true:
+     *
+     * <ul>
+     * <li>The class declaring the filter annotation is itself annotated with {@code @ApplicationScoped}</li>
+     * <li>The class contains instance fields that are not annotated with {@code Inject}. The reason for this rule
+     * is to ensure that fields will not be instantiated at static-init time, which can cause problems during
+     * native image build, see <a href="https://github.com/quarkusio/quarkus/issues/27752">this</a>
+     * for more details</li>
+     * </ul>
+     */
+    private boolean needsApplicationScoped(DotName dotName) {
+        ClassInfo classInfo = index.getClassByName(dotName);
+        if (classInfo == null) {
+            throw new IllegalStateException("Unable to find class '" + dotName + "' in index");
+        }
+        if (classInfo.hasDeclaredAnnotation(ApplicationScoped.class)) {
+            return true;
+        }
+        if (classInfo.hasDeclaredAnnotation(Singleton.class)) {
+            return false;
+        }
+        List<FieldInfo> fields = classInfo.fields();
+        if (fields.isEmpty()) {
+            return false;
+        } else {
+            boolean hasFieldWithoutInject = false;
+            for (FieldInfo field : fields) {
+                if (!field.hasAnnotation(Inject.class)) {
+                    hasFieldWithoutInject = true;
+                    break;
+                }
+            }
+            return hasFieldWithoutInject;
```

Comment:
What about `@RequestScoped`? Can users use this annotation? If so, you will add the `ApplicationScoped` if there are fields in spite of they annotated it with `@RequestScoped`.

Reply:
Yeah, I thought about that but was too lazy to take it into account. I guess now I have to :)